### PR TITLE
experimental: add descendent component

### PR DESCRIPTION
**Pull request by TrySound**
Ref https://github.com/webstudio-is/webstudio/issues/2903

Here added support for new `<ws:descendent />` component. It allows to style any children or descendents of components. Primary usage is for embed content which renders html and need to style tags.

Descendent component accepts "selector" prop which can be suffix like ` img`, ` .image` or `::first-child` pseudo-element.

Potentially it could be used as tag styling for any element.

Can be tested in any element by dropping "Descendent" component from "internal" category.

### CODE REVIEW

**Review by kof — APPROVED** [review]

File: apps/builder/app/canvas/shared/styles.ts [hunk 1]
```diff
@@ -159,6 +163,42 @@ const toVarValue = (styleDecl: StyleDecl): undefined | VarValue => {
   }
 };
 
+const $descendentSelectors = computed(
```
Comment (kof) [inline apps/builder/app/canvas/shared/styles.ts:166]: this def. needs tests

File: apps/builder/app/canvas/shared/styles.ts [hunk 2]
```diff
@@ -227,10 +267,31 @@ export const subscribeStyles = () => {
     }
   );
 
+  const unsubscribeDescendentSelectors = $descendentSelectors.subscribe(
```
Comment (kof) [inline apps/builder/app/canvas/shared/styles.ts:270]: def. needs tests

File: packages/css-engine/src/core/atomic.test.ts [hunk 3]
```diff
@@ -175,3 +175,33 @@ test("distinct similar declarations from different breakpoints", () => {
...
+  const sheet = createRegularStyleSheet();
+  sheet.addMediaRule("x");
+  const rule1 = sheet.addNestingRule("instance");
+  rule1.setDeclaration({
+    breakpoint: "x",
+    selector: ":hover",
+    property: "display",
+    value: { type: "keyword", value: "block" },
+  });
+  const rule2 = sheet.addNestingRule("instance", " img");
```
Comment (kof) [inline packages/css-engine/src/core/atomic.test.ts:189]: are you sure you want to have space and such as part of the selector string and not as something completely separated?

File: packages/css-engine/src/core/style-sheet-regular.test.ts [hunk 4]
```diff
@@ -932,6 +932,35 @@ describe("Style Sheet Regular", () => {
     color: blue
   }
 }"
+`);
+  });
+
+  test("support descendent suffix", () => {
+    const sheet = createRegularStyleSheet();
+    sheet.addMediaRule("base", {});
+    const rule1 = sheet.addNestingRule(".instance");
```
Comment (kof) [inline packages/css-engine/src/core/style-sheet-regular.test.ts:941]: .instance or .element? technically you are talking about an element that is backing the instance, but in this case you are talking specifically about the element, instance is basically the component instance that contains that element, but you are not styling the instance

File: packages/css-engine/src/core/rules.ts [hunk 5]
```diff
@@ -80,13 +81,25 @@ export class NestingRule {
...
     this.#mixinRules = mixinRules;
   }
   getSelector() {
     return this.#selector;
   }
+  setSelector(selector: string) {
+    this.#selector = selector;
+    this.#cache.clear();
+  }
+  getDescendentSuffix() {
```
[the remark this answers is not in the document]
Comment (TrySound) [reply packages/css-engine/src/core/rules.ts:100]: Here no combinators `::placeholder`

File: packages/css-engine/src/core/style-sheet-regular.test.ts (hunk 4 above)
Comment (TrySound) [reply packages/css-engine/src/core/style-sheet-regular.test.ts:941]: Just a reference `[data-ws-id=instanceId]` -> `.instance`

File: packages/css-engine/src/core/atomic.test.ts (hunk 3 above)
Comment (TrySound) [reply packages/css-engine/src/core/atomic.test.ts:189]: Yep, string is enough

**Pull request merged**